opentracing-util/src/main/java/io/opentracing/util/GlobalTracer.java

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright 2016-2017 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.util;
+
+import io.opentracing.NoopTracerFactory;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.propagation.Format;
+
+import java.util.Iterator;
+import java.util.ServiceLoader;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Forwards all methods to another tracer that can be configured in one of two ways:
+ * <ol>
+ * <li>Explicitly, calling {@link #register(Tracer)} with a configured tracer, or:</li>
+ * <li>Automatically using the Java {@link ServiceLoader} SPI mechanism to load a {@link Tracer} from the classpath.</li>
+ * </ol>
+ * <p>
+ * When the tracer is needed it is lazily looked up using the following rules:
```

Comment:
@sjoerdtalsma I will respond in this thread.

I also think that GlobalTracer should be simplified. 

> @pavolloffay I am afraid I don't quite understand your remark. If you have concerns about an app crash, please file an issue on the original repo, because that sounds dangerous.

The crash is not related to `GlobalTracer` itself. This tracer allows to re-register tracer at any time during the application lifetime which can cause app crash/exceptions. It allows you to register zipkin tracer and then hawkular tracer and if the code is using GlobalTracer like this 

```java
GlobalTracer.get()
    .buildSpan("foo")
    .asChildOf(child)
    .start();
```
 it would cause problems because child can be from a different tracer (and spec does not deal with this corner case).

Maybe `GlobalTracer.register()` should be only one-time  operation.